### ISSUE TITLE
Provide API support along with cli

### ISSUE THREAD

**Issue by danielo515** [robertklep/node-benchr#7]
Hello again,

Your library really reduces the amount of work necessary for running benchmarks.
However, in order to make it even easier (just drop your code and run) and easily operable for newcomers I'm creating a set of helper scripts.
Being a node.js developer, creating scripts also in node is the easier way to make the experience rich. I'm making a small inquirer script that asks you which benchmark files you want to run. ~~Currently I'm forced to execute benchr as an executable file from node (making use of spawn, which is not very reliable).~~ It will be much easier and portable to be able to require benchr and provide it with the files to run. 
Currently I can require the runner this way: `require('benchr/lib/runner.js')`. I have to do this because your package.json is lacking a main property. Adding `benchr/lib/runner.js` as main file (keeping the bin as is) will ease this kind of usage.

Regards

**Comment by danielo515** [comment]
I have to correct myself, launching the runner from another node script (making use of the Runner class) does not seem to work. Using spawn however, works as expected on MacOs

```js
 const spawn = require('child_process').spawn;
    const touch1 = spawn('npm', ['run', 'bench', '--', ...toRun ], { stdio: 'inherit' });
```

**Comment by robertklep** [comment]
Any idea why it didn't work? Because it's basically what `lib/cli.js` does.

The only thing you have to look out for is that you're passing the `options` argument with the proper defaults (provided by `docopt` in case of the CLI tool).

**Comment by danielo515** [comment]
> Any idea why it didn't work? Because it's basically what lib/cli.js does.

I had no idea until I read your next paragraph 

> The only thing you have to look out for is that you're passing the options argument with the proper defaults (provided by docopt in case of the CLI tool).

Do you know which are those defaults ? Reading the docopt documentation it is not obvious. (docopt looks much nicer than yargs, by the way)

**Comment by robertklep** [comment]
The defaults are actually inferred from the `docopt.txt` file (which is really very cool IMO, but a bit confusing if you're never used it):

```
{ '--delay': '0',
  '--grep': null,
  '--help': false,
  '--json': false,
  '--max-time': '5',
  '--min-time': '0',
  '--version': false,
  '<file>': [ 'test/bench' ] }
```

**Comment by danielo515** [comment]
May it be something like this ?

```
{
  "--delay": "0", 
  "--grep": null, 
  "--help": false, 
  "--max-time": "5", 
  "--min-time": "0", 
  "--version": false, 
  "<file>": [
    "benchr", 
    "filename"
  ]
}
``

**Comment by robertklep** [comment]
Haha yup!

**Comment by danielo515** [comment]
Just for your information, seems that the minimun options required for the  Runner class are:

```js
        const runOptions = {
            '--delay': '0',
            '--max-time': '5',
            '--min-time': '0',
        }
```
 If you are going to take advantage of ES6 features you may assign those defaults on the class, so when somebody uses just the class doesn't have to provide such options

```js
class Runner(options = { '--delay': '0', '--max-time': '5', '--min-time': '0' }, files )
```

And if you are going to separate the CLI from the class you may want to make some translation on the cli side so you can take a more DX :

```js
class Runner(options = { delay: 0, maxTime: 5, minTime: 0 }, files )
```

**Comment by robertklep** [comment]
I wanted to keep it compatible with Node v6, so no default arguments for now.

In case you want to test, check out the `json-output` branch: https://github.com/robertklep/node-benchr/tree/json-output

It's not yet finished, I'd like to add the option to pass a custom reporter module, but it has two reporters built-in now: `console` (the old one, and default) and `json` (JSON output).

There's an event-based reporter API which is pretty self-explanatory. Feel free to take a look at it (it's used by both built-in reporters) and let me know if you think it's acceptable. In the end, it should be possible for you to create a custom HTML-outputting reporter module, without having to use the JSON reporter.

**Comment by danielo515** [comment]
Hello @robertklep , that looks amazing.

I was about asking you how to specify the output, but I can see that it is the STDOUT.
I just run one benchmark and  the output is overwhelming. I can't tell you if it is acceptable or not because I have to figure out what each field means.

However, some things that I miss could be:
* Ability to specify a name to the test suite. Having the file name as default is nice, but it will be nicer to have a more human friendly (and potential label for HTML output).
* Ability to specify which fields to output. The report is 2k lines, which is splendidly verbose. Sure it is valuable informations, but I feel that some fields (like options) can be omitted.
* Still missing the main parameter on the package.json (probably you want to fix this later, just a reminder)

Taking a look at the code I can see that you are putting too much pressure on the Runner class. Surely it is hard to accept a custom reporter from console, but taking it as an optional argument (falling back to console) on the Runner class will ease the task of using custom reporters from other nodeJs scripts and will effectively decouple the runner from the cli.

I'll take a look at the console reporter to see if I can make any sense from the fields the  json one is creating.

Looking very nice so far.

**Comment by danielo515** [comment]
Taking a look at json reporter, specifically this part

```js
.on('suite.complete', suite => {
    verbose && process.stderr.write('s');
    currentSuite.benchmarks = suite.map(b => b);
  })
```

I ask myself: what if the user does not specify any suite ?  will she get an empty output ? Currently adding suites is optional, right?

**Comment by robertklep** [comment]
I agree it's a bit overwhelming.

Here's an abbreviated version that, I believe, contains all the data (and a bit more) that the `console` reporter uses:
```
[
  {
    "name": "test/bench",
    "suites": [
      {
        "name": "My Synchronous Benchmarks",
        "benchmarks": [
          {
            "name": "Regexp#test",
            "hz": 19719225.300944813,
            "aborted": false,
            "stats": {
              "moe": 2.449672149482792e-09,
              "rme": 4.830563702910093,
              "sem": 1.1704119204408943e-09,
              "deviation": 5.234241231563831e-09,
              "mean": 5.071193136335263e-08,
              "variance": 2.739728127020285e-17
            }
          },
          {
            "name": "String#indexOf",
            "hz": 534877395.2669684,
            "aborted": false,
            "stats": {
              "moe": 2.523709643516387e-11,
              "rme": 1.3498752405341745,
              "sem": 1.20983204387171e-11,
              "deviation": 5.544146919225246e-11,
              "mean": 1.8695873275797703e-09,
              "variance": 3.0737565061954783e-21
            }
          },
          ...
```

FWIW, suites are already named (`"name": "My Synchronous Benchmarks"`). You can always opt to not output the filename, or to take the filename and generate a name based on the path, when you're implementing a custom reporter.

**Comment by robertklep** [comment]
And yes, suites are optional, in which case they are given a generated name (`Suite#1`, `Suite#2`, ...). They are still created internally.

If you want to explicitly name them, use the `suite()` function:
```
suite('My Suite', function() {
  benchmark('Benchmark 1', function() {
    ...
  });

  benchmark('Benchmark 2', function() {
    ...
  });
});
```

**Comment by danielo515** [comment]
Yes, sorry for speking too soon, I just saw 
```js
 // Implicitly add a suite when there aren't any yet while defining this benchmark.
    if (this.suites.length === 0) {
      this.addSuite(() => {});
    }
```

That output that you posted is a manual edit or are you working on a slimmed down version on the JSON reporter ?

Are you working on the reporter right now, right ?  I don't want to toy with something that is currently WIP.

>  suites are already named ("name": "My Synchronous Benchmarks").

Yes, I'm already using it. Sorry, I wanted to express something different. I was thinking about a third level of grouping `fileNAme => Suite => benchmakr` but  I stupidly wrote just suite.

About the JSON output, I'm having several ideas but I am not sure if I like any of them.
My target is not only being able to show stats, but also to detect regressions and that stuff. 
For this matter, having the timestamp of each run will probably help. Maybe instead of just hardcoding the timestamp it will be easier to accept some metadata that will be embedded on the suite. One cool thing of this feature is that it could allow you to embed the node version and compare the performance of the different solutions on the different node versions.
On this regard I don't know if it will be better to create a new file on each run or just extend the same file adding new elements to the array.
You have probably more experience on this.

 What about the decoupling proposal ? Do you like it ?

**Comment by robertklep** [comment]
> That output that you posted is a manual edit or are you working on a slimmed down version on the JSON reporter ?

That's a proposal for a more compact output. Perhaps I could use the `--verbose` flag to enable outputting all data, and default to a more compact output.

> About the JSON output, I'm having several ideas but I am not sure if I like any of them.
My target is not only being able to show stats, but also to detect regressions and that stuff.

I think this is much better handled with a custom reporter, which I intend to implement before doing a new release. That way, you're completely free to control the output and the data it contains (even write everything to a database, if you'd like).

As for decoupling, that's something that seems to make sense. I want to decouple `docopt` from the `Runner` class completely, too.

**Comment by danielo515** [comment]
> I think this is much better handled with a custom reporter, which I intend to implement before doing a new release. That way, you're completely free to control the output and the data it contains (even write everything to a database, if you'd like).

You're absolutely right.

> As for decoupling, that's something that seems to make sense. I want to decouple docopt from the Runner class completely, too.

Glad to hear that. The next release might be awesome!!

**Comment by danielo515** [comment]
Hello @robertklep, 

I am already taking advantage of the JSON reporter. I am using C3 charts and ejs templates to build a nice looking dashboard with graphs for each benchmark file. It is looking very good so far.


![captura de pantalla 2018-01-28 a las 21 07 09](https://user-images.githubusercontent.com/2270425/35486514-5556b794-046f-11e8-8d01-22ce2b2b06e5.png)


The only thing that I'm missing so far (along with a smaller output 😄  ) that is already available on the console output is the fastest method of each suite. It should be an easy addition to the current JSON output right ?

**Comment by robertklep** [comment]
Yeah, should not be a problem to add. Really cool charts!

**Comment by danielo515** [comment]
Don't want to be overwhelming,  but I just added fastest support on the JSON reporter this way:

```js
}).on('suite.complete', suite => {
    verbose && process.stderr.write('s');
    currentSuite.benchmarks = suite.map(b => b);
    currentSuite.fastest = suite.filter('fastest').map('name');
  }).on('benchmark.start', () => {
```

It's working fine for me.
I won't disturb you any-more today.

Thanks again, and regards

**Comment by robertklep** [comment]
Take another look at the [`json-output`](https://github.com/robertklep/node-benchr/tree/json-output) branch.

The latest commit adds CLI/Runner decoupling, custom reporter modules, and various new options.

**Comment by danielo515** [comment]
I just checked your branch.
Looks awesome !! 
I made a couple of comments on your latest comment, hope you don't mind.

Something that I didn't mentioned before neither on the commit is that there is an undocumented feature. Well, it is not strictly undocumented, but I find it a bit hidden, specially on the master branch. According to the examples is quite obvious that you can pass a glob pattern to your CMD. What not many people knows is that such glob will be expanded to a list of files, so you can pass also a list of files separated by spaces. This is very convenient in several situations, for example in may case, where I collect a list of files to execute through a wizard.

Thanks again for all the enhancements you have implemented so fast. I have several projects and I know how hard is to be up-to-date with the users requests.

**Comment by danielo515** [comment]
Additionally will you considering upgrading to the latest Benchmark.js version? It is 2.1.2 while you are on 2.1.0

**Comment by robertklep** [comment]
> Additionally will you considering upgrading to the latest Benchmark.js version? It is 2.1.2 while you are on 2.1.0

Done! Also the other dependencies.

Check out the latest commit (7111414d84a1eb13165b19ce9b5a27a48d88fe1c), I think it's close to a releasable version 😄

**Comment by danielo515** [comment]
Sorry for being so active and for reporting everything on the same issue, if you want I can leave this issue as It is and open a new one just for the json reporter. 
In any case I think now is the right moment to report this because you are actively working on it. I just saw that Benchmark.js collects environment information:

https://benchmarkjs.com/docs#platform

This makes little sense for the console output, but it is of great interest for the json reporter. I think that including the object at the root of the report without any further modification is more than enough

**Comment by robertklep** [comment]
I'm happy to include it, although it would mean that the JSON output has to change from an array (of files of suites of benchmarks) to an object.

**Comment by danielo515** [comment]
> it would mean that the JSON output has to change from an array (of files of suites of benchmarks) to an object.

Mmm, not sure about that. What about including that information on every file object ? It is not much overhead and will help keeping the information isolated but accessible. Imagine that you want to run the test suites on different environments but store the information on the same file for easier comparision (which happens to be my case 😄 ). Storing the results on an array structure is not only easier to iterate, but also keeps original order. 
My use case is to create a framework for easing the environment setup ( by including a common set of libraries, tools and instructions) and running the same set of benchmarks across several versions of node and create a gallery with the results i n form of charts. Since it will be stored and hosted at github a single Json file is a perfect f it.
My plan was to use it with my co-workers , but nothing prevents it from going wild as a community tool.

Regards

**Comment by robertklep** [comment]
> What about including that information on every file object?

Seems a bit superfluous. I was thinking about something like this:
```
{
  "timestamp" : "...",
  "platform" : { ... },
  "files" : [ ... ]
}
```

Where `files` is what's now being outputted (still in order, and equally easy to iterate/compare).

**Comment by danielo515** [comment]
Hello again @robertklep 

What you propose makes a lot of sense. Maybe I was being a bit selfish (and lazy 😄 ).
However, I still to see weird that each report includes all the files. To me, each js file containing a suite or set of suites should be a separate entity. I see several cases where I will only execute a small set of files (as small as one).
In any case, you can provide the information aggregated and I can de-normalize it, but for that to happen I would need a different reporting mechanism than console, as I said.

Regards

**Comment by robertklep** [comment]
I understand that each file can be regarded as a separate entity, but there's only a single piece of output that gets generated for each entire run (which may consist of multiple files).

Even if the built-in JSON reporter would accept a function argument to pass the output _to_, it still would call that function at the end of a full run, and the output object would have to contain all the data that I mentioned (`timestamp`, `platform`, `files`).

If you want separate outputs for each file, a custom reporter can generate the output in the `file.complete` event handler instead of in the `run.complete` handler.

**Comment by danielo515** [comment]
Hello @robertklep ,

Again, what you says makes a lot of sense.
I think the easier and best way would be to handle that custom logic (if needed) on the output receiver function, it should not be hard.

I'm very expectant about the release so I can implement my tool with the full set of features we are talking about.

**Comment by robertklep** [comment]
I just published v4.0.0

**Comment by danielo515** [comment]
I'm trying it out right now!

**Comment by danielo515** [comment]
Here are some things that It may worth mentioning:

- Because you are using docop, you are getting safe defaults for free. However, progragmatic usage can get tricky if the correct parameters are not provided. For example, if you omit `maxTime, minTime` or `delay` nothing will break, but the test will run indefinitely. 
- Accepting the reporter as function makes it easier to provide custom reporters. However, if you want to use a custom one we are on the opposite situation, you are forced to "break" your packages encapsulation and check for the reporters path. In mu case I am forced to do `const reporter = require(`benchr/lib/reporters/${outputType}`);`. So now I have a cleaner runner import (`const Runner = require('benchr')`) but I still need to traverse your internal folder structure. It will be nice to accept both methods
- The size of the reports and their format has improved dramatically. Very well done!
- The reporting mechanism is still a bit unconvenient. I think we can discuss this topic on a separate issue. Meanwhile I'm just cheating a bit: 
```js
if (outputType === 'json') console.log = (format, stats) => writeObj(makePath('../reports/report.json'), stats);
new Runner(runOptions, toRun).run();
```

Regards

**Comment by robertklep** [comment]
> Because you are using docop, you are getting safe defaults for free. However, progragmatic usage can get tricky if the correct parameters are not provided. For example, if you omit maxTime, minTime or delay nothing will break, but the test will run indefinitely.

Oops, I forgot to set those defaults in the runner. Will fix ASAP.

> Accepting the reporter as function makes it easier to provide custom reporters. However, if you want to use a custom one we are on the opposite situation, you are forced to "break" your packages encapsulation and check for the reporters path. In mu case I am forced to do `const reporter = require(benchr/lib/reporters/${outputType})`

I think we have different ideas on what a reporter does: for me, they are modules that both gather the benchmarking results and output these results (in some form or another).

The built-in reporters aren't intended to be re-used, or "overloaded" somehow. They are there to provide two default forms of output, for users that aren't interested in anything else.

If you _do_ want something else (different format, write the output to a file/database, etc), you should write a custom reporter. Or use the built-in JSON reporter and use some shell scripting glue (The UNIX Way).

**Comment by danielo515** [comment]
> I think we have different ideas on what a reporter does: for me, they are modules that both gather the benchmarking results and output these results (in some form or another).

No, we have the same idea. I understand reporters exactly as you described.

> The built-in reporters aren't intended to be re-used, or "overloaded" somehow. They are there to provide two default forms of output, for users that aren't interested in anything else.

I will not say that I would find useful to be able to overload the default reporters, but I was not complaining about that. What I want to express is that there is no easy programmatic way of using the default reporters. In my case I created a small wizard using inquirer that allows you to select which benchmarks to run and what reporter to use. Once I gathered the answers I just want to launch the runner with the provided answers, which is a list of files and one of the built-in reporters. Since you are not accepting a string anymore I am forced to manually require the built-in reporters if I want to use any of them.
Hope  I expressed myself clearer.

Regards

**Comment by danielo515** [comment]
I'll close this issue because the latest release provides what I was asking for and a bit more.
Very good job, and an amazing communication flow. Thank you very much.

Regards

**Comment by robertklep** [comment]
> Since you are not accepting a string anymore I am forced to manually require the built-in reporters if I want to use any of them. Hope I expressed myself clearer.

Yes, absolutely, and now I understand the problem.

If you think it's acceptable to pass either a string or a function, it should be an easy fix.

**Comment by danielo515** [comment]
> If you think it's acceptable to pass either a string or a function, it should be an easy fix.

Well, on an ideal stricter world, probably not (maybe using an options object with named options or something like that).  On a pragmatic or real world scenario,  I think it should be acceptable 😄

**Comment by robertklep** [comment]
I think that the reporter resolving code that's now done in `lib/cli.js` should be moved to `lib/runner.js`, so it's possible to pass either a string (which is _either_ the name of a built-in reporter _or_ the name of a file) or a function.

**Comment by robertklep** [comment]
Just released v4.2.0
